oauth_tools/external_idp.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import abc
+import logging
+from os.path import join
+from time import sleep
+from typing import List, Optional
+
+import requests
+from lightkube import Client, KubeConfig, codecs
+from lightkube.core.exceptions import ApiError, ObjectDeleted
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Namespace, Pod, Service
+from pytest_operator.plugin import OpsTest
+from requests.exceptions import RequestException
+
+from oauth_tools.constants import (
+    DEX_CLIENT_ID,
+    DEX_CLIENT_SECRET,
+    DEX_MANIFESTS,
+    EXTERNAL_USER_EMAIL,
+    EXTERNAL_USER_PASSWORD,
+    KUBECONFIG,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalIdpManager(abc.ABC):
+    """Abstract class for managing lifecycle for an external IdP."""
+
+    @property
+    def client_id(self) -> str:
```

Comment:
Following this case, is it better to explicitly make it `abstractmethod`?

Reply:
Meant to add it and forgot about it. Fixed